src/include/buffer/buffer_pool_manager_instance.h

```diff
@@ -0,0 +1,171 @@
+//===----------------------------------------------------------------------===//
+//
+//                         BusTub
+//
+// buffer_pool_manager_instance.h
+//
+// Identification: src/include/buffer/buffer_pool_manager.h
+//
+// Copyright (c) 2015-2021, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <list>
+#include <mutex>  // NOLINT
+#include <unordered_map>
+
+#include "buffer/buffer_pool_manager.h"
+#include "buffer/lru_replacer.h"
+#include "recovery/log_manager.h"
+#include "storage/disk/disk_manager.h"
+#include "storage/page/page.h"
+
+namespace bustub {
+
+/**
+ * BufferPoolManager reads disk pages to and from its internal buffer pool.
+ */
+class BufferPoolManagerInstance : public BufferPoolManager {
+ public:
+
+  /**
+   * Creates a new BufferPoolManagerInstance.
+   * @param pool_size the size of the buffer pool
+   * @param disk_manager the disk manager
+   * @param log_manager the log manager (for testing only: nullptr = disable logging)
+   */
+  BufferPoolManagerInstance(size_t pool_size, DiskManager *disk_manager, LogManager *log_manager = nullptr);
+
+  /**
+   * Destroys an existing BufferPoolManagerInstance.
+   */
+  ~BufferPoolManagerInstance() override;
+
+  /** Grading function. Do not modify! */
+  Page *FetchPage(page_id_t page_id, bufferpool_callback_fn callback = nullptr) override {
+    GradingCallback(callback, CallbackType::BEFORE, page_id);
```

Comment:
It's probably possible to push these functions into the parent class and make the *Impl functions virtual. That way it's not duplicated across the sub-classes. Thoughts?